File: frontend/src/types/timeline.ts
/**
 * Timeline TypeScript types for Evidence Timeline feature
 * Feature: 002-evidence-timeline
 */

export type TimelineEventType =
    | 'message'
    | 'document'
    | 'image'
    | 'audio'
    | 'video'
    | 'incident';

export interface TimelineEvent {
    eventId: string;
    evidenceId: string;
    caseId: string;

    // Time
    timestamp: string | null;  // ISO 8601 or null for unknown dates
    date: string;              // YYYY-MM-DD or "날짜 미상"
    time: string;              // HH:MM or ""

    // Content
    description: string;
    contentPreview?: string;

    // Classification
    eventType: TimelineEventType;
    labels: string[];
    speaker?: string;          // 원고/피고/제3자/unknown
    sourceFile: string;

    // Significance
    significance: number;      // 1-5 scale
    isKeyEvidence: boolean;    // significance >= 4

    // Metadata
    metadata?: Record<string, unknown>;
}

export interface TimelineFilter {
    dateStart?: string;        // YYYY-MM-DD
    dateEnd?: string;          // YYYY-MM-DD
    labels?: string[];
    speakers?: string[];
    eventTypes?: TimelineEventType[];
    keyOnly?: boolean;

    // Pagination
    limit?: number;            // 1-100, default 50
    offset?: number;           // default 0

    // Sort
    sortOrder?: 'asc' | 'desc'; // default 'asc' (oldest first)
}

export interface TimelineResult {
    caseId: string;
    events: TimelineEvent[];

    // Pagination
    totalCount: number;
    filteredCount: number;
    hasMore: boolean;

    // Statistics
    dateRange: {
        start: string | null;
        end: string | null;
    };
    eventsByType: Record<string, number>;
    eventsByLabel: Record<string, number>;
    keyEventsCount: number;

    // Meta
    generatedAt: string;       // ISO 8601
}

// Helper type for filter state management
export interface TimelineFilterState extends TimelineFilter {
    isFiltering: boolean;
}

// Speaker options for filter dropdown
export const SPEAKER_OPTIONS = ['원고', '피고', '제3자', 'unknown'] as const;
export type SpeakerOption = typeof SPEAKER_OPTIONS[number];

// Common Article 840 labels for filter dropdown
export const COMMON_LABELS = [
    '부정행위',
    '폭행',
    '학대',
    '유기',
    '협박',
    '위협',
    '폭언',
    '계속적_불화',
    '혼인_파탄',
    '재산_문제',
    '양육_문제',
] as const;

File: backend/app/schemas/timeline.py
"""
Timeline Schemas
Pydantic models for evidence timeline feature

Feature: 002-evidence-timeline
"""

from datetime import datetime
from enum import Enum
from typing import Dict, List, Optional

from pydantic import BaseModel, Field


class TimelineEventType(str, Enum):
    """Evidence type enum for timeline events"""
    MESSAGE = "message"
    DOCUMENT = "document"
    IMAGE = "image"
    AUDIO = "audio"
    VIDEO = "video"
    INCIDENT = "incident"


class TimelineEvent(BaseModel):
    """Individual timeline event representing a piece of evidence"""
    event_id: str = Field(..., description="Unique event identifier")
    evidence_id: str = Field(..., description="Reference to evidence record")
    case_id: str = Field(..., description="Parent case ID")
    timestamp: Optional[datetime] = Field(None, description="Event timestamp")
    date: str = Field(..., description="Formatted date string (YYYY-MM-DD)")
    time: str = Field(..., description="Formatted time string (HH:MM)")
    description: str = Field(..., description="Event description/summary")
    content_preview: Optional[str] = Field(None, description="Preview of content")
    event_type: TimelineEventType = Field(..., description="Type of evidence")
    labels: List[str] = Field(default_factory=list, description="Article 840 labels")
    speaker: Optional[str] = Field(None, description="Speaker/source name")
    source_file: str = Field(..., description="Original file name")
    significance: int = Field(ge=1, le=5, default=1, description="Significance score 1-5")
    is_key_evidence: bool = Field(False, description="Whether this is key evidence")
    metadata: Optional[Dict] = Field(None, description="Additional metadata")

    class Config:
        from_attributes = True


class TimelineFilter(BaseModel):
    """Filter criteria for timeline queries"""
    date_start: Optional[str] = Field(None, description="Start date (YYYY-MM-DD)")
    date_end: Optional[str] = Field(None, description="End date (YYYY-MM-DD)")
    labels: Optional[List[str]] = Field(None, description="Filter by labels")
    speakers: Optional[List[str]] = Field(None, description="Filter by speakers")
    event_types: Optional[List[TimelineEventType]] = Field(None, description="Filter by event types")
    key_only: bool = Field(False, description="Show only key evidence")
    limit: int = Field(50, ge=1, le=200, description="Max results per page")
    offset: int = Field(0, ge=0, description="Pagination offset")
    sort_order: str = Field("asc", description="Sort order: asc or desc")


class DateRange(BaseModel):
    """Date range for timeline data"""
    start: Optional[str] = None
    end: Optional[str] = None


class TimelineResult(BaseModel):
    """Timeline query result with events and statistics"""
    case_id: str = Field(..., description="Case ID")
    events: List[TimelineEvent] = Field(default_factory=list, description="Timeline events")
    total_count: int = Field(0, description="Total events count")
    filtered_count: int = Field(0, description="Filtered events count")
    has_more: bool = Field(False, description="Whether more results exist")
    date_range: DateRange = Field(default_factory=DateRange, description="Date range of events")
    events_by_type: Dict[str, int] = Field(default_factory=dict, description="Event count by type")
    events_by_label: Dict[str, int] = Field(default_factory=dict, description="Event count by label")
    key_events_count: int = Field(0, description="Number of key evidence items")
    generated_at: str = Field(..., description="Timestamp when result was generated")

    class Config:
        from_attributes = True
